R/c14_date_list_calibrate.R

```diff
@@ -143,3 +102,102 @@ calibrate.c14_date_list <- function(x) {
 
   return(x)
 }
+
+#' determine_dates_out_of_range_of_calcurve
+#'
+#' @param x c14_date_list
+#'
+#' @return indizes of dates in c14_date_list which are out of calcurve range
+determine_dates_out_of_range_of_calcurve <- function(x) {
+  # load intcal13 data from Bchron
+  utils::data("intcal13", package = "Bchron")
+
+  toona <- which(is.na(x$c14age) | is.na(x$c14std))
+  toosmall <- which(x$c14age < min(intcal13[,2]))
+  toobig <- which(x$c14age > max(intcal13[,2]))
+  outofrange <- c(toona, toosmall, toobig) %>% unique
+
+  return(outofrange)
+}
+
+#' calibrate_to_probability_distribution
+#'
+#' @param x c14_date_list
+#' @param ... passed to Bchron::BchronCalibrate()
+#'
+#' @return list with probability distribution data frames
+calibrate_to_probability_distribution <- function(x, ...) {
+  Bchron::BchronCalibrate(
+    ages      = x$c14age,
+    ageSds    = x$c14std,
+    calCurves = rep("intcal13", nrow(x)),
+    ...
+  ) %>%
+    # extract probability distribution from BchronCalibrate output
+    # and craft new data.frame with year and density
+    lapply(
+      function(x) {
+        data.frame(calage = x[["ageGrid"]], density = x[["densities"]])
+      }
+    ) %>%
+    return()
+}
+
+#' determine_calage_from_probability_distribution
+#'
+#' @param x c14_date_list
+#'
+#' @return vector of calages
+determine_calage_from_probability_distribution <- function(x) {
+  x %>%
+    lapply(
+      function(x) {
+        x$calage[which.max(x$density)]
+      }
+    ) %>%
+    unlist %>%
+    return()
+}
+
+#' Calculate highest density regions for Bchron calibrated ages
+#'
+#' A function for computing highest density regions (HDRs). Adopted from the Bchron package
+#'
+#' @details The output of this function is a list of contiguous ranges which cover the probability interval requested. A highest density region might have multiple such ranges if the calibrated date is multi-modal. These differ from credible intervals, which are always contiguous but will not be a good representation of a multi-modal probability distribution.
+#'
+#' @param calprobdistr the probability distribution
+#' @param prob The desired probability interval, in the range(0, 1)
+#'
+#' @return a dataframe containing the hdr
+hdr <- function(calprobdistr, prob = 0.95) {
```

Comment:
Should this function be more atomized? Small subfunctions for the individual steps?

Reply:
Don't think so, not necessary. It is a rather straight forward, procedural function. But if anyone else like to do that...